File: telly/android/core/src/main/kotlin/uk/telly/core/M3u.kt
package uk.telly.core

/** One playable channel, as the playlist describes it. */
data class Channel(
    val id: String,
    val number: Int,
    val name: String,
    val url: String,
    val logo: String,
    val group: String,
    val tvgId: String
) {
    val key: String get() = name.lowercase()
}

/** What a parsed playlist carries besides its channels. */
data class Playlist(
    val channels: List<Channel>,
    val epgUrl: String
)

object M3u {

    private val ATTR = Regex("""([A-Za-z0-9_-]+)\s*=\s*"([^"]*)"""")
    private val URI = Regex("""^[a-zA-Z][a-zA-Z0-9+.-]*://""")

    private fun attrs(s: String): Map<String, String> =
        ATTR.findAll(s).associate { it.groupValues[1].lowercase() to it.groupValues[2] }

    /**
     * Split an #EXTINF payload into its attributes and its title, without
     * tripping over commas inside attribute values or channel names.
     */
    internal fun splitExtinf(payload: String): Pair<String, String> {
        var i = 0
        while (i < payload.length && (payload[i].isDigit() || payload[i] == '-' || payload[i] == '.')) i++
        var inQuotes = false
        while (i < payload.length) {
            val c = payload[i]
            if (c == '"') inQuotes = !inQuotes
            else if (c == ',' && !inQuotes) break
            i++
        }
        val attrText = payload.substring(0, minOf(i, payload.length))
        val title = if (i < payload.length) payload.substring(i + 1).trim() else ""
        return attrText to title
    }

    private fun safeName(url: String): String =
        url.substringBefore('?').substringAfterLast('/').ifEmpty { url }

    fun parse(text: String): Playlist {
        val channels = ArrayList<Channel>()
        val seen = HashSet<String>()
        var epgUrl = ""
        var pendingName = ""
        var pendingLogo = ""
        var pendingGroup = ""
        var pendingTvgId = ""
        var hasPending = false
        var extgrp = ""

        for (raw in text.lineSequence()) {
            val line = raw.trim()
            if (line.isEmpty()) continue

            if (line.startsWith("#EXTM3U")) {
                val a = attrs(line)
                epgUrl = a["url-tvg"] ?: a["x-tvg-url"] ?: ""
                continue
            }
            if (line.startsWith("#EXTINF:")) {
                val (attrText, title) = splitExtinf(line.substring(8))
                val a = attrs(attrText)
                pendingName = title.ifEmpty { a["tvg-name"] ?: "Unnamed channel" }
                pendingTvgId = a["tvg-id"] ?: ""
                pendingLogo = a["tvg-logo"] ?: ""
                pendingGroup = a["group-title"] ?: ""
                hasPending = true
                continue
            }
            if (line.startsWith("#EXTGRP:")) { extgrp = line.substring(8).trim(); continue }
            if (line.startsWith("#")) continue                 // #EXTVLCOPT, comments

            // A line that is not a URI is not a stream: real playlists carry
            // placeholders such as [NO PUBLIC STREAM] for channels with no feed.
            if (!URI.containsMatchIn(line)) {
                hasPending = false; extgrp = ""
                continue
            }

            val name = if (hasPending) pendingName else safeName(line)
            val group = (if (hasPending) pendingGroup else "").ifEmpty { extgrp.ifEmpty { "Ungrouped" } }
            val id = "$group|$line"
            if (seen.add(id)) {
                channels.add(
                    Channel(
                        id = id,
                        number = channels.size + 1,
                        name = name,
                        url = line,
                        logo = if (hasPending) pendingLogo else "",
                        group = group,
                        tvgId = if (hasPending) pendingTvgId else ""
                    )
                )
            }
            hasPending = false; extgrp = ""
        }
        return Playlist(channels, epgUrl)
    }

    /** Ordered category names, in the order the playlist introduces them. */
    fun groups(channels: List<Channel>): List<String> =
        LinkedHashSet(channels.map { it.group }).toList()
}

File: telly/server/src/services/m3u.js
/**
 * The M3U parser, moved server-side.
 *
 * This is the same parser the web and Android clients use, edge case for edge
 * case: commas inside quoted attributes and inside channel names, #EXTGRP,
 * #EXTVLCOPT, addresses that are placeholders rather than URIs, duplicate
 * streams within a group, stable channel numbers.
 */
const ATTR = /([A-Za-z0-9_-]+)\s*=\s*"([^"]*)"/g;
const URI = /^[a-zA-Z][a-zA-Z0-9+.-]*:\/\//;

function attrs(s) {
  const out = {};
  let m;
  ATTR.lastIndex = 0;
  while ((m = ATTR.exec(s))) out[m[1].toLowerCase()] = m[2];
  return out;
}

/** Split an #EXTINF payload into attributes and title without tripping on commas. */
export function splitExtinf(payload) {
  let i = 0;
  while (i < payload.length && /[-\d.]/.test(payload[i])) i++;
  let inQuotes = false;
  for (; i < payload.length; i++) {
    const c = payload[i];
    if (c === '"') inQuotes = !inQuotes;
    else if (c === ',' && !inQuotes) break;
  }
  return { attrText: payload.slice(0, i), title: payload.slice(i + 1).trim() };
}

function safeName(url) {
  try { return decodeURIComponent(url.split('?')[0].split('/').pop()) || url; }
  catch { return url; }
}

/**
 * Movies and series are guessed from the group title, the same way the clients
 * do it today, so nothing changes for an existing playlist. When a source
 * carries proper VOD (Xtream does) the loader sets `kind` explicitly instead.
 */
export function classify(group) {
  if (/serie|season|episode|tv ?show/i.test(group)) return 'series';
  if (/movie|film|vod|cinema/i.test(group)) return 'movie';
  return 'live';
}

export function parseM3u(text) {
  const channels = [];
  const seen = new Set();
  let epgUrl = '';
  let pending = null;
  let extgrp = '';

  for (const raw of String(text).split(/\r?\n/)) {
    const line = raw.trim();
    if (!line) continue;

    if (line.startsWith('#EXTM3U')) {
      const a = attrs(line);
      epgUrl = a['url-tvg'] || a['x-tvg-url'] || '';
      continue;
    }
    if (line.startsWith('#EXTINF:')) {
      const { attrText, title } = splitExtinf(line.slice(8));
      const a = attrs(attrText);
      pending = {
        name: title || a['tvg-name'] || 'Unnamed channel',
        tvgId: a['tvg-id'] || '',
        logo: a['tvg-logo'] || '',
        group: a['group-title'] || ''
      };
      continue;
    }
    if (line.startsWith('#EXTGRP:')) { extgrp = line.slice(8).trim(); continue; }
    if (line.startsWith('#')) continue;
    if (!URI.test(line)) { pending = null; extgrp = ''; continue; }

    const meta = pending || { name: safeName(line), tvgId: '', logo: '', group: '' };
    const group = meta.group || extgrp || 'Ungrouped';
    const extId = `${group}|${line}`;
    if (!seen.has(extId)) {
      seen.add(extId);
      channels.push({
        extId,
        number: channels.length + 1,
        name: meta.name,
        group,
        logo: meta.logo,
        tvgId: meta.tvgId,
        url: line,
        kind: classify(group)
      });
    }
    pending = null; extgrp = '';
  }
  return { channels, epgUrl };
}
